igvc_navigation/src/swerve_drive/swerve_drive.cpp

```diff
@@ -0,0 +1,240 @@
+#include <parameter_assertions/assertions.h>
+
+#include "swerve_drive.h"
+
+SwerveDrive::SwerveDrive() : pNh{ "~" }
+{
+  if (!getParams())
+  {
+    ROS_ERROR_STREAM("Unable to import parameters!");
+    return;
+  }
+  assertions::param(pNh, "max_vel", max_vel_, 3.0);
+  mbf_twist_ = nh.subscribe("/cmd_vel", 1, &SwerveDrive::twistToVelocity, this);
+  vel_pub_ = nh.advertise<igvc_msgs::velocity_quad>("/motors", 1);
+}
+
+void SwerveDrive::twistToVelocity(geometry_msgs::Twist twist)
+{
+  if (!std::isfinite(twist.linear.x) || !std::isfinite(twist.linear.y) || !std::isfinite(twist.angular.z))
+  {
+    ROS_WARN_THROTTLE(1.0, "Recieved NaN in velocity command. Ignoring.");
+    return;
+  }
+
+  const double speedX = twist.linear.x;
+  const double speedY = twist.linear.y;
+  const double rotation = twist.angular.z;
+
+  double localMax = INT_MIN;
+
+  const size_t num_wheels = 4;
+  for (size_t i = 0; i < num_wheels; ++i)
+  {
+    double wheel_vx = speedX - rotation * positions_list[i][1];
+    double wheel_vy = speedY + rotation * positions_list[i][0];
+
+    double wheel_V = sqrt(pow(wheel_vx, 2) + pow(wheel_vy, 2));
+    double wheel_dir = atan2(wheel_vy, wheel_vx);
+
+    int omega_dir = set_command_angle(wheel_dir, i);
+
+    localMax = std::max(localMax, wheel_V);
+
+    wheel_info[i][0] = wheel_V * omega_dir;
+  }
+
+  double normalization = 1.0;
+  if (localMax > max_vel_)
+  {
+    normalization = max_vel_ / localMax;
+  }
+  // contructing vel message
+  igvc_msgs::velocity_quad vel_msg;
+  vel_msg.fl_velocity = wheel_info[0][0] * normalization;
+  vel_msg.bl_velocity = wheel_info[1][0] * normalization;
+  vel_msg.br_velocity = wheel_info[2][0] * normalization;
+  vel_msg.fr_velocity = wheel_info[3][0] * normalization;
+
+  vel_msg.fl_angle = wheel_info[0][1];
+  vel_msg.bl_angle = wheel_info[1][1];
+  vel_msg.br_angle = wheel_info[2][1];
+  vel_msg.fr_angle = wheel_info[3][1];
+  vel_msg.duration = 0.02;
+  vel_msg.header.stamp = ros::Time::now();
+
+  vel_pub_.publish(vel_msg);
+}
+
+int SwerveDrive::set_command_angle(const double& target, const int wheel_idx)
```

Comment:
Nit, but probably change `const int wheel_idx` to `const int& wheel_idx` for consistency